File: middleware-moshi/src/test/java/com/nytimes/android/external/store3/middleware/moshi/MoshiSourceParserTest.java
package com.nytimes.android.external.store3.middleware.moshi;

import com.nytimes.android.external.store3.base.Fetcher;
import com.nytimes.android.external.store3.base.Parser;
import com.nytimes.android.external.store3.base.Persister;
import com.nytimes.android.external.store3.base.impl.BarCode;
import com.nytimes.android.external.store3.base.impl.ParsingStoreBuilder;
import com.nytimes.android.external.store3.base.impl.Store;
import com.nytimes.android.external.store3.middleware.moshi.data.Foo;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.io.ByteArrayInputStream;
import java.nio.charset.Charset;

import io.reactivex.Maybe;
import io.reactivex.Single;
import okio.BufferedSource;
import okio.Okio;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class MoshiSourceParserTest {

    private static final String KEY = "key";
    private static final String sourceString =
            "{\"number\":123,\"string\":\"abc\",\"bars\":[{\"string\":\"def\"},{\"string\":\"ghi\"}]}";
    @Rule
    public ExpectedException expectedException = ExpectedException.none();
    @Mock
    Fetcher<BufferedSource, BarCode> fetcher;
    @Mock
    Persister<BufferedSource, BarCode> persister;
    private final BarCode barCode = new BarCode("value", KEY);

    private static BufferedSource source(String data) {
        return Okio.buffer(Okio.source(new ByteArrayInputStream(data.getBytes(Charset.defaultCharset()))));
    }

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);

        BufferedSource bufferedSource = source(sourceString);
        assertNotNull(bufferedSource);

        when(fetcher.fetch(barCode))
                .thenReturn(Single.just(bufferedSource));

        when(persister.read(barCode))
                .thenReturn(Maybe.<BufferedSource>empty())
                .thenReturn(Maybe.just(bufferedSource));

        when(persister.write(barCode, bufferedSource))
                .thenReturn(Single.just(true));
    }

    @Test
    public void testSourceParser() throws Exception {

        Parser<BufferedSource, Foo> parser = MoshiParserFactory.createSourceParser(Foo.class);

        Store<Foo, BarCode> store = ParsingStoreBuilder.<BufferedSource, Foo>builder()
                .persister(persister)
                .fetcher(fetcher)
                .parser(parser)
                .open();

        Foo result = store.get(barCode).blockingGet();

        assertEquals(result.number, 123);
        assertEquals(result.string, "abc");
        assertEquals(result.bars.size(), 2);
        assertEquals(result.bars.get(0).string, "def");
        assertEquals(result.bars.get(1).string, "ghi");

        verify(fetcher, times(1)).fetch(barCode);

    }

    @Test
    public void testNullMoshi() {
        expectedException.expect(NullPointerException.class);
        MoshiParserFactory.createSourceParser(null, Foo.class);
    }

    @Test
    public void testNullType() {
        expectedException.expect(NullPointerException.class);
        MoshiParserFactory.createSourceParser(null, Foo.class);
    }

}

File: middleware-moshi/src/test/java/com/nytimes/android/external/store3/middleware/moshi/MoshiSourceParserTest.kt
package com.nytimes.android.external.store3.middleware.moshi

import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import com.nytimes.android.external.store3.base.Fetcher
import com.nytimes.android.external.store3.base.Persister
import com.nytimes.android.external.store3.base.impl.BarCode
import com.nytimes.android.external.store3.base.impl.ParsingStoreBuilder
import kotlinx.coroutines.runBlocking
import okio.BufferedSource
import okio.Okio
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Rule
import org.junit.Test
import org.junit.rules.ExpectedException
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import java.io.ByteArrayInputStream
import java.nio.charset.Charset

class MoshiSourceParserTest {
    @Rule
    @JvmField
    var expectedException = ExpectedException.none()
    private val fetcher: Fetcher<BufferedSource, BarCode> = mock()
    private val persister: Persister<BufferedSource, BarCode> = mock()
    private val barCode = BarCode("value", KEY)

    fun setUp() = runBlocking<Unit> {
        val bufferedSource = source(sourceString)
        assertNotNull(bufferedSource)

        whenever(fetcher.fetch(barCode))
                .thenReturn(bufferedSource)

        whenever(persister.read(barCode))
                .thenReturn(null)
                .thenReturn(bufferedSource)

        whenever(persister.write(barCode, bufferedSource))
                .thenReturn(true)
    }

    @Test
    fun testSourceParser() = runBlocking<Unit> {

        val parser = MoshiParserFactory.createSourceParser<Foo>(Foo::class.java)

        val store = ParsingStoreBuilder.builder<BufferedSource, Foo>()
                .persister(persister)
                .fetcher(fetcher)
                .parser(parser)
                .open()

        val result = store.get(barCode)

        assertEquals(result.number.toLong(), 123)
        assertEquals(result.string, "abc")
        assertEquals(result.bars.size.toLong(), 2)
        assertEquals(result.bars[0].string, "def")
        assertEquals(result.bars[1].string, "ghi")

        verify<Fetcher<BufferedSource, BarCode>>(fetcher, times(1)).fetch(barCode)

    }

    @Test
    fun testNullMoshi() {
        expectedException.expect(NullPointerException::class.java)
        MoshiParserFactory.createSourceParser<Any>(null!!, Foo::class.java)
    }

    @Test
    fun testNullType() {
        expectedException.expect(NullPointerException::class.java)
        MoshiParserFactory.createSourceParser<Any>(null!!, Foo::class.java)
    }

    companion object {

        private val KEY = "key"
        private val sourceString = "{\"number\":123,\"string\":\"abc\",\"bars\":[{\"string\":\"def\"},{\"string\":\"ghi\"}]}"

        private fun source(data: String): BufferedSource {
            return Okio.buffer(Okio.source(ByteArrayInputStream(data.toByteArray(Charset.defaultCharset()))))
        }
    }

}
